### PR TITLE
High Level Review

### PR DESCRIPTION
**Pull request by agoessling**

### CODE REVIEW

File: test/test_c_stuff_sack.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "external/unity/src/unity.h"
```
Comment (ghorn) [inline test/test_c_stuff_sack.c:6]: You should #include external stuff with <> instead of "". See my comment in the BUILD file for unity.

File: tools/unity.BUILD [hunk 2]
```diff
@@ -0,0 +1,9 @@
+cc_library(
```
Comment (ghorn) [inline tools/unity.BUILD:1]: See
https://docs.bazel.build/versions/main/be/c-cpp.html#cc_library.includes
https://docs.bazel.build/versions/main/be/c-cpp.html#cc_library.include_prefix
https://docs.bazel.build/versions/main/be/c-cpp.html#cc_library.strip_include_prefix
and mess around so that you can include unity with <>. This is important because bazel will treat it as a system dependency and not subject it to the many compiler warnings that you will want to add.

I would recommend `#include <third_party/unity/unity.h>`

Comment (ghorn) [reply tools/unity.BUILD:1]: Also consider moving this from `tools` to a new top-level `third_party` directory.

File: src/c_stuff_sack.py [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+    s += _indent('{};\n'.format(struct_field_declaration(field)))
+
+  s += '}} {};'.format(struct.name)
+
+  return s
+
+
+def pack(type_object):
+  if isinstance(type_object, ss.Bitfield):
+    return primitive_union_pack(type_object)
```
Comment (ghorn) [inline src/c_stuff_sack.py:128]: Can you add a comment why this is handled by union_pack instead of primitive_pack?

File: src/c_stuff_sack.py [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+
+  if isinstance(type_object, ss.Primitive):
+    return primitive_pack(type_object)
+
+  raise TypeError('Unknown type: {}'.format(type(type_object)))
+
+
+def unpack(type_object):
+  if isinstance(type_object, ss.Bitfield):
+    return primitive_union_unpack(type_object)
```
Comment (ghorn) [inline src/c_stuff_sack.py:150]: Can you add a comment why this is handled by union_unpack instead of primitive_unpack?

File: test/test_message_spec.yaml [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    - boolean: bool
+    - float_type: float
+    - double_type: double
+
+ArrayElem:
+  type: Struct
+  fields:
+    - field0: bool
+    - field1: uint16
+
```
Comment (ghorn) [inline test/test_message_spec.yaml:320]: Consider testing and empty struct.

**Review by ghorn — COMMENTED** [review]
I read all of the code, though I skimmed the sphynx generation and didn't try to poke any holes in the test code. I like the design choice of not generating any python except for a shim - it means that you can read and lint real python code. That python is a bit hard to follow. It would be easier if I could have started by reading some generated C/python code with the sphynx docs, hosted on github.

I'm impressed that you also made a bazel macro to tightly integrate the spec parsing, code generation, and bazel packaging. Consider documenting that it can be used as a stand-alone tool as well.

Host on github an end to end example starting with a yaml spec of a few structs, enums, bitfields, and an array. Host the generated C/python and sphynx docs. It would make this a lot more accessible.

Explain in the README or in a Design Choices doc:
* why enums are signed
* tradeoffs in adding the two fields to enums (polluting switch case statements)
* why the two extra enum fields are dropped from python
* Why enums are forced to be the smallest size - what gcc flag is needed to support this on x86
* multidimensional arrays always being C-style [x][y][z]
* empty structs allowed? not allowed? related: compatibility with C++
* Endianness assumptions
* C language standard support if there are any restrictions (I'd worry about empty structs/bitfields, bools, empty arrays maybe, and just lack of testing of simple things which could be easily ported to older C standards)

Why 3 levels of bazel dependencies? Is this so that a user doesn't have to get the sphynx deps if they don't want? Have you tested using this project from another bazel project?

You may be interested in the bazel BUILD file formatter https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md

I strongly recommend using multiline f-strings for as much of this as you possibly can. It will make the generator code look a lot more like the generated code which really helps readability.

Consider in your test code, compiling the generated code with a comprehensive set of C warnings/errors. You could start with -Weverything and remove false positive flags until you are left with real warnings, if you want. You should add corner cases like empty structs and enums to your tests.

I think you would get a huge amount of testing value from property tests https://hypothesis.readthedocs.io/en/latest/. I would generate random messages, generate and compile the C code (catching one class of errors), and then load the generated python and round trip many random python instances of each message and check that messages round-trip correctly (catching another class of errors). I found a lot of corner cases doing this in a similar library I once wrote. I would expect to find meaningful errors both in the compilation and the runtime stages.

Consider a test case where the spec defines an illegal mutually recursive struct like:

```
typedef struct {
  Bar bar;
} Foo;

typedef struct {
  Foo foo;
} Bar;
```

if it causes an infinite loop, consider figuring out how to raise an assertion instead. I ran into this and found it hard to debug in a similar library I once wrote.

Do you know if this library supports two message specs at the same time? The only issue that occurred to me was the mutation of DataType.all_types, though you could plausibly have related issues with closures.

I didn't check - are you using a wrapper around bitfields so that you can reference fields by index or name, and also display as both? Are you using a wrapper around enums as well? This would be good to document in a top-level README as example usability.

overall my main comments:
* property tests would bring the most value if you want to harden this
* avoid globals for readability
* host end-to-end example including sphynx docs
